### DISCUSSION TITLE
Heisenbug with fopen() in unit test

### DISCUSSION THREAD

**Discussion by Emil808** [nasa/fprime#1430] [Q&A]
I am unit testing a component that would access a text file and read one line to get an IP address and port number to configure a TCP Client socket 
The set up procedure would include, setting private variable `char hostfile_name[256]` to the text file, opening the file and saving the file pointer as a private member of the component implementation class. Then iterate through an array of TCP Client sockets, it will run a function to read one line from the file, then configure the socket, and attempt opening the connection.

Here is a snippet of the code. (This would be run at component initialization). 
```
this->fp = fopen(this->hostfile_name, "r"); 
if(this->fp == NULL){
(void)fprintf(stderr,"fopen on %s returned NULL ERRNO: %d\n", this->hostfile_name, errno);
        FW_ASSERT(this->fp); 
}
```
When running with `fprime-util check`, fopen() returns null, and triggers the FW_ASSERT with ERRNO = 2 (No such file or directory). 

But when running with the gdb debugger, either stepping through, or setting a break point past this area, fopen() returns a valid fp and the code continues. 

Here is a a sample of how the files are organized

```
Drv
  |- MultiTcpServer
  |   |- test/ut
  |   |- Tester.cpp
  |   |- Tester.hpp
  |   |- TestMain.cpp
  |   |  |- file
  |   |  |  |- hostfile_name_1.txt
  |- MultiTcpServerComponentImpl.hpp
  |- MultiTcpServerComponentImpl.cpp
  |- Task.hpp
  |- Task.cpp
  |- . . . and all other hpp/cpp helper files for this component
  ```
The Task class has a pure virtual function `getHostfileEntry(char *)` that expects the char* to be filled with a line read from the file. The ComponentImpl class will have `char hostfile_name[]` and `FILE *fp` that will be filled and set up at the component level set up. 

Is there any problems with using fopen to read a file in during unit test that is causing this failure when running `fprime-util check`? 
When putting this in a topology, and running a deployment, where should files be? 
How should the file path be? In Tester.cpp, the hostfile_name is set up as `test/ut/file/hostfile_name_1.txt` 
How should the filename/filepath be when I want it in the deployment? 
What are possible reasons that running check results in a 'no such file or directory' errno, but running in debugger it works fine?

**Comment by Emil808** [comment]
Yay okay! fprime-util check would need the file path starting from root. I had it originally as relative to where the Tester.cpp is. Changing it to the full file path to the file fixed it. 

Then what is the proper explanation for this behavior?

**Reply by LeStarch** [reply]
fpp-check runs the tests within the build cache not the local folder.  Thus paths are relative to somewhere in there.
